lib/active_model/serializer/adapter/json_api.rb

```diff
@@ -51,10 +77,11 @@ def serializable_hash_for_collection(serializer, options)
         end
 
         def serializable_hash_for_single_resource(serializer, options)
+          hash[:data] = {}
           primary_data = primary_data_for(serializer, options)
           relationships = relationships_for(serializer)
           included = included_for(serializer)
-          hash = { data: primary_data }
+          hash[:data].merge!(primary_data)
```

Comment:
On a technical note, it's a mutation solely because you assign it to `{}` right before. Maybe a bit of context will be helpful: the main goal of the JsonApi refactor I did was to decouple methods as much as possible so that they don't modify their arguments and just return the value they're supposed to compute. Hence, there is no need to either initialize `hash[:data]` or `merge` anything into it.